modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EVariable.java

```diff
@@ -43,27 +45,52 @@ public EVariable(Location location, String name) {
 
     @Override
     Output analyze(ClassNode classNode, ScriptRoot scriptRoot, Scope scope, Input input) {
-        if (input.read == false && input.write == false) {
-            throw createError(new IllegalArgumentException("not a statement: variable [" + name + "] not used"));
-        }
-
         Output output = new Output();
+        Class<?> type = scriptRoot.getPainlessLookup().canonicalTypeNameToType(name);
 
-        Variable variable = scope.getVariable(location, name);
+        if (type != null)  {
+            if (input.write) {
+                throw createError(new IllegalArgumentException("invalid assignment: " +
+                        "cannot write a value to a static type [" + PainlessLookupUtility.typeToCanonicalTypeName(type) + "]"));
+            }
 
-        if (input.write && variable.isFinal()) {
-            throw createError(new IllegalArgumentException("Variable [" + variable.getName() + "] is read-only."));
-        }
+            if (input.read == false) {
+                throw createError(new IllegalArgumentException("not a statement: " +
+                        "static type [" + PainlessLookupUtility.typeToCanonicalTypeName(type) + "] not used"));
+            }
+
+            output.actual = type;
+            output.isStaticType = true;
+
+            StaticNode staticNode = new StaticNode();
 
-        output.actual = variable.getType();
+            staticNode.setLocation(location);
+            staticNode.setExpressionType(output.actual);
 
-        VariableNode variableNode = new VariableNode();
+            output.expressionNode = staticNode;
+        } else if (scope.isVariableDefined(name)) {
```

Comment:
Say you have a variable with same as package name, eg `java`, what happens?

Reply:
eg `def java = 1; java.lang.Integer`

Reply:
This `int java; java.util.Integer.parseInt('1')` results in an error since the java package name is overridden by a variable name.

This `int java; java.util.ArrayList x = new java.util.ArrayList()` is successful as our parser can recognize that declarations must be a type, so it uses a different parsing path based on the rules.

This `int java; List x = new java.util.ArrayList(); ArrayList y; y = (java.util.ArrayList)x;` is successful for the same reasons as the previous example.

These all follow similar behavior to Java with error messages differing based on our use cases.